# Agent Customization & Prompt Engineering in AutoGen v0.4

In [1]:
from autogen_agentchat.agents import AssistantAgent
from autogen_ext.models.openai import OpenAIChatCompletionClient
from dotenv import load_dotenv
import os

load_dotenv()
# api_key = os.getenv('OPENAI_API_KEY')
groq_model_client = OpenAIChatCompletionClient(
    model="llama-3.3-70b-versatile",
    api_key=os.getenv("GROQ_API_KEY"),
    base_url="https://api.groq.com/openai/v1",
    model_info={
        "family": "llama",
        "vision": False,
        "function_calling": True,
        "json_output": True,
        "structured_output": True,
    },
)


### Agent Customization

- We can assign a role to our agent
- help in fitting agent to specific use case


In [2]:
asssistant = AssistantAgent(
    name = 'history_expert',
    model_client=groq_model_client,
    description='A knowledgeable assistant with expertise in world history',
    system_message='You are a history expert with deep knowledge of world history. Provide detailed and accuragte answers about historical events,figures and timelines'
)

In [3]:
async def test_history_expert():
    result = await asssistant.run(task = 'Who was the first President of USA?')
    print(result.messages[-1].content)
    

In [4]:
await test_history_expert()

The first President of the United States of America was George Washington (1732-1799). He was a military leader, statesman, and founding father who played a key role in the American Revolutionary War and the drafting of the United States Constitution.

Washington was unanimously elected as the first President of the United States by the Electoral College in 1789, and he served two terms in office from April 30, 1789, to March 4, 1797. He was inaugurated on the steps of Federal Hall in New York City, which was the temporary capital of the United States at the time.

As the first President, Washington established many of the customs and traditions that would shape the office of the presidency, including the decision to serve only two terms, which would later become the 22nd Amendment to the Constitution. He also established the first presidential cabinet, which included the Secretary of State, Secretary of War, Secretary of the Treasury, and Attorney General.

Washington's leadership and

The first President of the United States was George Washington (1732-1799). He was a military leader, statesman, and Founding Father who played a key role in the American Revolutionary War and the drafting of the United States Constitution.

George Washington was unanimously elected as the first President of the United States by the Electoral College on April 30, 1789. He was inaugurated on April 30, 1789, and served two terms in office until March 4, 1797.

Washington's presidency set important precedents for the executive branch of the federal government, and he is often referred to as the "Father of Our Country" for his role in shaping the United States. He established the cabinet system, defined the role of the president, and issued the Proclamation of Neutrality, which declared that the United States would remain neutral in foreign conflicts.

Some of Washington's notable achievements during his presidency include:

1. Establishing the Supreme Court and the federal judiciary system
2. Creating the Cabinet system, with the appointment of the Secretary of State, Secretary of War, Secretary of the Treasury, and Attorney General
3. Issuing the Proclamation of Neutrality, which declared that the United States would remain neutral in foreign conflicts
4. Suppressing the Whiskey Rebellion, a tax protest in western Pennsylvania
5. Signing the Jay Treaty, which resolved outstanding issues with Great Britain

Washington's legacy has endured, and he is widely regarded as one of the most important figures in American history. He is remembered for his integrity, leadership, and commitment to the principles of liberty and democracy.

The first President of the United States was George Washington. He served from 1789 to 1797. Before his presidency, Washington was well-known for his role in the American Revolutionary War as the commander-in-chief of the Continental Army. He is also noted for his presidency because he set many of the precedents for the office that are still followed today, such as limiting himself to only two terms in office. His presidency is significant for establishing numerous structures and procedures that still exist in the U.S. government today.

# Prompt Engineering

[Prompt Engineering - OpenAI ](https://platform.openai.com/docs/guides/prompt-engineering)

In [5]:
asssistant = AssistantAgent(
    name = 'history_expert',
    model_client=groq_model_client,
    description='A knowledgeable assistant with expertise in world history',
    system_message='You are a history expert with deep knowledge of world history who explaines event in a storytelling style, as if narrating a historic novel. Be vivid and engaging and focus on figures and events'
)

In [6]:
async def test_storytelling():
    result = await asssistant.run(task = 'Tell me about something on the signing of Declaration of Independence')
    print(result.messages[-1].content)
    

In [7]:
await test_storytelling()

It was a sweltering summer day in Philadelphia, July 4, 1776. The air was thick with anticipation as the Continental Congress gathered in the Assembly Room of the Pennsylvania State House, now known as Independence Hall. The weight of history hung in the balance, as fifty-six brave men prepared to sign their names to a document that would change the course of American history forever: the Declaration of Independence.

As the morning sun streamed through the windows, illuminating the faces of the assembled delegates, Thomas Jefferson, the primary author of the document, stood nervously, awaiting the moment of truth. He had poured his heart and soul into crafting the Declaration, with its eloquent phrases and bold assertions of American rights and freedoms. Now, it was time for the other delegates to put their signatures to the document, and with it, their lives, fortunes, and sacred honor on the line.

John Hancock, the president of the Continental Congress, a man of bold spirit and str

In [12]:
import asyncio
from autogen_agentchat.agents import AssistantAgent
from autogen_ext.models.openai import OpenAIChatCompletionClient
from dotenv import load_dotenv
import os

# Load API key from .env
load_dotenv()
api_key = os.getenv("OPENAI_API_KEY")
model_client = OpenAIChatCompletionClient(
    model="llama-3.3-70b-versatile",
    api_key=os.getenv("GROQ_API_KEY"),
    base_url="https://api.groq.com/openai/v1",
    model_info={
        "family": "llama",
        "vision": False,
        "function_calling": True,
        "json_output": True,
        "structured_output": True,
    },
)

# Define a simple tool
def get_weather(city: str) -> str:
    return f"The weather in {city} is sunny!"  # Dummy response

# Set up the agent
agent = AssistantAgent(
    name="weather_agent",
    model_client=model_client,
    system_message="You are a weather assistant. Use the get_weather tool when asked about weather.",
    tools=[get_weather],
    reflect_on_tool_use=True
)

# Run the agent
async def main():
    result = await agent.run(task="What’s the weather in New York?")
    print(result.messages[-1].content)

await main()


The current weather in New York is sunny. Would you like more details, such as the temperature or forecast?
